src/upchunk.spec.ts

```diff
@@ -1,333 +1,298 @@
-  let ATTEMPT_FAILURE_COUNT = 0;
-  const FAILURES = 2;
+  it('a single chunk failing the max number of times fails the upload', (done) => {
+    xhrMock.put(
+      endpoint,
+      sequence([
+        { status: 502 },
+        { status: 502 },
+        { status: 502 },
+        { status: 502 },
+        { status: 502 },
+        { status: 200 },
+      ])
+    );
 
-  nock('https://example.com')
-    .put('/upload/endpoint')
-    .times(FAILURES)
-    .reply(502)
-    .put('/upload/endpoint')
-    .twice()
-    .reply(200);
+    const upload = createUploadFixture({ delayBeforeAttempt: 0.1 });
 
-  const upload = createUploadFixture({ delayBeforeAttempt: 0.1 });
+    upload.on('error', (err) => {
+      try {
+        expect(err.detail.chunk).to.equal(0);
+        expect(err.detail.attempts).to.equal(5);
+        done();
+      } catch (err) {
+        done(err);
+      }
+    });
 
-  upload.on('attemptFailure', (err) => {
-    ATTEMPT_FAILURE_COUNT += 1;
+    upload.on('success', () => {
+      done(new Error(`Expected upload to fail due to failed attempts`));
+    });
   });
 
-  upload.on('error', done);
+  it('chunkSuccess event is fired after each successful upload', (done) => {
+    let calls = 0;
+    xhrMock.put(endpoint, { status: 200 });
 
-  upload.on('success', () => {
-    if (ATTEMPT_FAILURE_COUNT === FAILURES) {
-      return done();
-    }
+    const upload = createUploadFixture();
 
-    done(
-      `Expected ${FAILURES} attempt failures, received ${ATTEMPT_FAILURE_COUNT}`
-    );
-  });
-});
+    upload.on('chunkSuccess', () => {
+      calls = calls + 1;
+    });
 
-test('a single chunk failing the max number of times fails the upload', (done) => {
-  nock('https://example.com')
-    .put('/upload/endpoint')
-    .times(5)
-    .reply(502)
-    .put('/upload/endpoint')
-    .twice()
-    .reply(200);
-
-  const upload = createUploadFixture({ delayBeforeAttempt: 0.1 });
-
-  upload.on('error', (err) => {
-    try {
-      expect(err.detail.chunk).toBe(0);
-      expect(err.detail.attempts).toBe(5);
+    upload.on('success', () => {
+      expect(calls).to.equal(2);
       done();
-    } catch (err) {
-      done(err);
-    }
-  });
-
-  upload.on('success', () => {
-    done(`Expected upload to fail due to failed attempts`);
+    });
   });
-});
-
-test('chunkSuccess event is fired after each successful upload', (done) => {
-  nock('https://example.com')
-    .put('/upload/endpoint')
-    .reply(200)
-    .put('/upload/endpoint')
-    .reply(200);
 
-  const upload = createUploadFixture();
+  const isNumberArraySorted = (a: number[]): boolean => {
+    for (let i = 0; i < a.length - 1; i += 1) {
+      if (a[i] > a[i + 1]) {
+        return false;
+      }
+    }
+    return true;
+  };
+
+  it('progress event fires the correct upload percentage', (done) => {
+    xhrMock.put(endpoint, { status: 200 });
+    const fileBytes = 1048576;
+    const upload = createUploadFixture(
+      {
+        headers: {
```

Comment:
Here's the at least slightly dirty way of getting progress events working for xhr-mock in the same conditions as the previous test implementation using nock.